README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# pcal: Calibration of P-Values for Point Null Hypotheses

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/pcal)](https://CRAN.R-project.org/package=pcal)
[![Build
Status](https://travis-ci.org/pedro-teles-fonseca/pcal.svg?branch=master)](https://travis-ci.org/pedro-teles-fonseca/pcal)
[![R build
status](https://github.com/pedro-teles-fonseca/pcal/workflows/R-CMD-check/badge.svg)](https://github.com/pedro-teles-fonseca/pcal/actions)
![pkgdown](https://github.com/pedro-teles-fonseca/pcal/workflows/pkgdown/badge.svg)
[![codecov](https://codecov.io/gh/pedro-teles-fonseca/pcal/branch/master/graph/badge.svg)](https://codecov.io/gh/pedro-teles-fonseca/pcal)
[![MIT
license](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://lbesson.mit-license.org/)
<!-- badges: end -->

## Overview

P-values are the most commonly used tool to measure the evidence
provided by the data against a model or hypothesis. Unfortunately,
p-values are often incorrectly interpreted as the probability that the
null hypothesis is true or as type I error probabilities. The `pcal`
package uses the calibrations developed by Sellke, Bayarri, and Berger
(2001) to calibrate p-values under a robust perspective and obtain
measures of the evidence provided by the data in favor of point null
hypotheses which are safer and more straightforward interpret:

  - `pcal()` calibrates p-values so that they can be directly
    interpreted as either lower bounds on the posterior probabilities of
    point null hypotheses or as lower bounds on type I error
    probabilities. With this calibration one need not fear the
    misinterpretation of a type I error probability as the probability
    that the null hypothesis is true because they coincide. Note that
    the output of this calibration has both Bayesian and Frequentist
    interpretations.

  - `bcal()` calibrates p-values so that they can be interpreted as
    lower bounds on Bayes factors in favor of point null hypotheses.

Some utility functions are also included:

  - `bfactor_to_prob()` turns Bayes factors into posterior probabilities
    using a formula from Berger and Delampady (1987).

  - `bfactor_interpret()` classifies the strength of the evidence
    implied by a Bayes factor according to the scales suggested by
    Jeffreys (1961) and Kass and Raftery (1995).

  - `bfactor_log_interpret()` is similar to `bfactor_interpret()` but
    takes logarithms of Bayes factors as input.

## Installation

The released version of `pcal` can be installed from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("pcal")
```

The development version can be installed from
[GitHub](https://github.com/) using the `devtools` package:

``` r
# install.packages("devtools")
devtools::install_github("pedro-teles-fonseca/pcal")
```

## Usage

First we need a p-value from any statistical test of a point null
hypothesis:

``` r
x <- matrix(c(22, 13, 13, 23), ncol = 2)
pv <- chisq.test(x)[["p.value"]]

pv
#> [1] 0.04377308
```

In classical hypothesis testing, if the typical 0.05 significance
threshold is used then this p-value slightly below 0.05 would result in
the rejection of the null hypothesis.

With `bcal()` we can turn `pv` into a lower bound for the Bayes factor
in favor of the null hypothesis:

``` r
bcal(pv)
#> [1] 0.3722807
```

We can also turn `pv` into a lower bound for the posterior probability
of the null hypothesis using `pcal()`:

``` r
pcal(pv)
#> [1] 0.2712861
```

This is an approximation to the minimum posterior probability of the
null hypothesis that we would find by changing the prior distribution of
the parameter of interest (under the alternative hypothesis) over wide
classes of distributions. The output of `bcal()` has an analogous
interpretation for Bayes factors (instead of posterior probabilities).

Note that according to `pcal()` the posterior probability that the null
hypothesis is true is at least 0.27 (approximately), which implies that
a p-value below 0.05 is not necessarily indicative of strong evidence
against the null hypothesis.

One can avoid the specification of prior probabilities for the
hypotheses by focusing solely on Bayes factors. To compute posterior
probabilities for the hypotheses, however, prior probabilities must by
specified. By default, `pcal()` assigns a prior probability of 0.5 to
the null hypothesis. We can specify different prior probabilities, for
example:

``` r
pcal(pv, prior_prob = .95)
#> [1] 0.8761354
```

In this case we obtain a higher lower bound because the null hypothesis
has a higher prior probability.

Sellke, Bayarri, and Berger (2001) noted that a scenario in which they
definitely recommend the aforementioned calibrations is when
investigating fit to the null model with no explicit alternative in
mind. Pericchi and Torres (2011) warned that despite the usefulness and
appropriateness of these p-value calibrations they do not depend on
sample size, and hence the lower bounds obtained with large samples may
be conservative.

Since the output of `bcal(pv)` is a Bayes factor, we can use
`bfactor_to_prob()` to turn it into a posterior probability:

``` r
bfactor_to_prob(bcal(pv)) # same as pcal(pv)
#> [1] 0.2712861
```

Like `pcal()`, `bfactor_to_prob()` assumes a prior probability of 0.5 to
the null hypothesis. We can change this default:

``` r
bfactor_to_prob(bcal(pv), prior_prob = .95)
#> [1] 0.8761354
```

To classify the strength of the evidence in favor of the null hypothesis
implied by a Bayes factor according to the scale suggested by Jeffreys
(1961) we can use `bfactor_interpret()`:

``` r
bfs <- c(0.1, 2, 5, 20, 50, 150)

bfactor_interpret(bfs)
#> [1] "Negative"    "Weak"        "Substantial" "Strong"      "Very Strong"
#> [6] "Decisive"
```

Alternatively, we can use the interpretation scale suggested by Kass and
Raftery (1995):

``` r
bfactor_interpret(bfs, scale = "kass-raftery")
#> [1] "Negative"    "Weak"        "Positive"    "Strong"      "Strong"     
#> [6] "Very Strong"
```

Because Bayes factors are often reported on a logarithmic scale, there
is also a `bfactor_log_interpret()` function that interprets the
logarithms of Bayes factors:

``` r
log_bfs <- log10(bfs)
 
bfactor_log_interpret(log_bfs, base = 10)
#> [1] "Negative"    "Weak"        "Substantial" "Strong"      "Very Strong"
#> [6] "Decisive"

bfactor_log_interpret(log_bfs, scale = "kass-raftery", base = 10)
#> [1] "Negative"    "Weak"        "Positive"    "Strong"      "Strong"     
#> [6] "Very Strong"
```

To compare Bayes factors with results from standard likelihood ratio
tests it can be useful to obtain the strength of the evidence against
the null hypothesis. If `bf` is a Bayes factor in favor of the null
hypothesis, one can use `1/bf` as input to obtain the strength of the
evidence against the null hypothesis:

``` r
# Evidence in favor of the null hypothesis
bfactor_interpret(c(0.1, 2, 5, 20, 50, 150))
#> [1] "Negative"    "Weak"        "Substantial" "Strong"      "Very Strong"
#> [6] "Decisive"

# Evidence against the null hypothesis
bfactor_interpret(1/c(0.1, 2, 5, 20, 50, 150))
#> [1] "Strong"   "Negative" "Negative" "Negative" "Negative" "Negative"
```

## Getting Help

If you find a bug, please file an issue with a minimal reproducible
example on [GitHub](https://github.com/pedro-teles-fonseca/pcal/issues).
Feature requests are also welcome. You can contact me at
<pedro.teles.fonseca@phd.iseg.ulisboa.pt>.

## References

<div id="refs" class="references hanging-indent">

<div id="ref-bergerDelampady1987">

Berger, James O., and Mohan Delampady. 1987. “Testing Precise
Hypotheses.” *Statistical Science* 2 (3): 317–35.

</div>

<div id="ref-jeffreys1961">

Jeffreys, Harold. 1961. *Theory of Probability*. 3rd ed. Oxford Classic
Texts in the Physical Sciences. Oxford University Press.

</div>

<div id="ref-kass1995">

Kass, Robert E., and Adrian E. Raftery. 1995. “Bayes Factors.” *Journal
of the American Statistical Association* 90 (430): 773–95.

</div>

<div id="ref-pericchiTorres2011">

Pericchi, Luis, and David Torres. 2011. “Quick Anomaly Detection by the
Newcomb—Benford Law, with Applications to Electoral Processes Data from
the USA, Puerto Rico and Venezuela.” *Statistical Science* 26 (4):
502–16.

</div>

<div id="ref-sellke2001">

Sellke, Thomas, M. J. Bayarri, and James O. Berger. 2001. “Calibration
of P Values for Testing Precise Null Hypotheses.” *The American
Statistician* 55 (1): 62–71.

</div>

</div>


NEWS.md

# pcal 1.1.0

- This is a minor release, and the second CRAN release of `pcal`. 

## Major changes 

* `bfactor_interpret()` has been deprecated in favor of `polya::bfactor_interpret()`.

* `bfactor_log_interpret()` has been deprecated in favor of `polya::bfactor_log_interpret()`.

* `bfactor_interpret_kr` has been deprecated. `bfactor_interpret(bf, scale = "kass-raftery")` should be used instead.

* `bfactor_log_interpret_kr` has been deprecated. `bfactor_log_interpret(bf, scale = "kass-raftery")` should be used instead.

* `bfactor_interpret()` and `bfactor_log_interpret()` now have a `scale` argument to choose the Bayes factor interpretation scale to be used. The default `scale` is `"Jeffreys"`, which means that `bfactor_interpret(bf)` gives the same output as `bfactor_interpret(bf, scale = "Jeffreys")`. With the default `scale` both `bfactor_interpret()` and `bfactor_log_interpret()`  give the same output as they did in `pcal` 1.0.0.

## Minor changes

* `package?pcal` is now documented.

* Error and warning messages are now more informative.

* Improved code efficiency. 

* Improved unit tests.

* Improved README.

* Improved documentation.

* Improved code readability.

* Licence changed from GPL 3.0 to MIT.

# pcal 1.0.0

This is the first CRAN release of `pcal`.








R/deprecated.R

# nocov start

#' @title Interpretation of Bayes factors
#'
#' @description Quantify the strength of the evidence provided by the data to a model/hypothesis according to Bayes factor interpretation scales suggested by \insertCite{jeffreys1961;textual}{pcal} and \insertCite{kass1995;textual}{pcal}.
#'
#' @param bf A numeric vector of non-negative values.
#' @param scale A string of characters specifying either "jeffreys" or "kass-raftery". Not case sensitive.
#'
#' @details Bayes factors are a summary of the evidence provided by the data to a model/hypothesis. \insertCite{jeffreys1961;textual}{pcal} suggested the interpretation of Bayes factors in half-units on the base 10 logarithmic scale, as indicated in the following table:
#'
#' **log10(Bayes factor)**  | **Bayes factor** | **Evidence** |
#' | ---                    | ---              | ---          |
#' | [-Inf, 0[              | [0, 1[           | Negative     |
#' | [0, 0.5[               | [1, 3.2[         | Weak         |
#' | [0.5, 1[               | [3.2, 10[        | Substantial  |
#' | [1, 1.5[               | [10, 32[         | Strong       |
#' | [1.5, 2[               | [32, 100[        | Very Strong  |
#' | [2, +Inf[              | [100, +Inf[      | Decisive     |
#'
#' By default, `bfactor_interpret` takes Bayes factors as input and returns the strength of the evidence in favor of the model/hypothesis in the numerator of the Bayes factors (usually the null hypothesis) according to the aforementioned table.
#'
#' Alternatively, and because it can be useful to consider twice the natural logarithm of the Bayes factor (which is in the same scale as the familiar deviance and likelihood ratio test statistics), \insertCite{kass1995;textual}{pcal} suggested the following scale:
#'
#' | **2*log(Bayes factor)** | **Bayes factor**      | **Evidence** |
#' | ---                     | ---                   | ---          |
#' | [-Inf, 0[               | [0, 1[                | Negative     |
#' | [0, 2[                  | [1, 3[                | Weak         |
#' | [2, 6[                  | [3, 20[               | Positive     |
#' | [6, 10[                 | [20, 150[             | Strong       |
#' | [10, +Inf[              | [150, +Inf[           | Very Strong  |
#'
#' To interpret Bayes factors according to the latter table use `scale = "kass-raftery"`.
#'
#' When comparing Bayes factors with results from standard likelihood ratio tests it is convenient to put the null hypothesis in the denominator of the Bayes factor so that `bfactor_interpret` returns the strength of the evidence against the null hypothesis. If `bf` was obtained with the null hypothesis on the numerator, one can use `bfactor_interpret(1/bf)` or `bfactor_interpret(1/bf, scale = "kass-raftery")` to obtain the strength of the evidence against the null hypothesis.
#'
#' @return Returns a character vector with the same {\link[base]{length}} as `bf`.
#'
#' @references
#' \insertAllCited{}
#'
#' @seealso
#' * \code{\link[pcal]{bfactor_log_interpret}} for the interpretation of the logarithms of Bayes factors.
#' * \code{\link[pcal]{bfactor_to_prob}} to turn Bayes factors into posterior probabilities.
#' * \code{\link[pcal]{bcal}} for a p-value calibration that returns lower bounds on Bayes factors in favor of point null hypotheses.
#'
#' @examples
#' # Interpretation of one Bayes factor
#' bfactor_interpret(4)
#' bfactor_interpret(4, scale = "jeffreys")
#' bfactor_interpret(4, scale = "kass-raftery")
#'
#' # Interpretation of more than one Bayes factors
#' bfactor_interpret(c(0.1, 1.2, 3.5, 13.9, 150))
#' bfactor_interpret(c(0.1, 1.2, 3.5, 13.9, 150), scale = "jeffreys")
#' bfactor_interpret(c(0.1, 1.2, 3.5, 13.9, 150), scale = "kass-raftery")
#'
#' # Application: chi-squared goodness-of-fit test.
#' # Strength of the evidence provided by the lower
#' # bound of the Bayes factor in favor of the null hypothesis:
#' data <- matrix(c(18, 12, 10, 12, 10, 23), ncol = 2)
#' bfactor_interpret(bcal(chisq.test(data)[["p.value"]]))
#'
#' @export

bfactor_interpret <- function(bf, scale = "jeffreys") {

  .Deprecated(
    new = "bfactor_interpret",
    package = "polya",
    msg = "bfactor_interpret' is deprecated. Please use
    'polya::bfactor_interpret(bf)'
    instead.")

  if (is.null(bf)){
    stop(paste("Invalid argument: bf is NULL."))
  }
  if (any(!is.numeric(bf), !is.vector(bf))){
    stop(paste("Invalid argument: bf must be a numeric vector."))
  }
  if (length(bf) == 0){
    stop(paste("Invalid argument: bf is empty."))
  }
  if (all(is.na(bf))){
    stop(paste("Invalid argument: all elements of bf are NA or NaN."))
  }
  if (any(bf[!is.na(bf)] < 0)){
    stop(paste("Invalid argument: all elements of bf must be non-negative."))
  }
  if (any(is.na(bf))){
    warning(paste("There are NA or NaN values in bf."))
  }

  if (is.null(scale)){
    stop(paste("Invalid argument: scale is NULL."))
  }
  if (any(!is.vector(scale), !is.atomic(scale), isFALSE(length(scale) == 1))){
    stop(paste("Invalid argument: scale must be an atomic vector of length 1."))
  }
  if (is.na(scale)){
    stop(paste("Invalid argument: scale is NA or NaN."))
  }
  if (!is.character(scale)){
    stop(paste("Invalid argument: the type of scale must be character."))
  }
  if (isFALSE(tolower(scale) %in% c("jeffreys", "kass-raftery"))){
    stop(paste("Invalid argument: scale must be either 'jeffreys' or 'kass-raftery'."))
  }

  switch(
    tolower(scale),
    "jeffreys" = ifelse(bf < 1, "Negative",
                        ifelse(bf < 3.2, "Weak",
                               ifelse(bf < 10, "Substantial",
                                      ifelse(bf < 32, "Strong",
                                             ifelse(bf < 100, "Very Strong",
                                                    "Decisive")
                                      )
                               )
                        )
    ),

    "kass-raftery" = ifelse(bf < 1, "Negative",
                            ifelse(bf < 3, "Weak",
                                   ifelse(bf < 20, "Positive",
                                          ifelse(bf < 150, "Strong",
                                                 "Very Strong")
                                   )
                            )
    )
  )
}

#' @title Interpretation of the logarithms of Bayes factors
#'
#' @description Quantify the strength of the evidence provided by the data to a model/hypothesis according to Bayes factor interpretation scales suggested by \insertCite{jeffreys1961;textual}{pcal} and \insertCite{kass1995;textual}{pcal}.
#'
#' @param bf A numeric vector.
#' @param scale A string of characters specifying either "jeffreys" or "kass-raftery". Not case sensitive.
#' @param base A numeric vector of \code{\link[base]{length}} one. Must be a positive number.
#'
#'
#' @details Bayes factors are a summary of the evidence provided by the data to a model/hypothesis, and are often reported on a logarithmic scale. \insertCite{jeffreys1961;textual}{pcal} suggested the interpretation of Bayes factors in half-units on the base 10 logarithmic scale, as indicated in the following table:
#'
#'
#' **log10(Bayes factor)**  | **Bayes factor** | **Evidence** |
#' | ---                    | ---              | ---          |
#' | [-Inf, 0[              | [0, 1[           | Negative     |
#' | [0, 0.5[               | [1, 3.2[         | Weak         |
#' | [0.5, 1[               | [3.2, 10[        | Substantial  |
#' | [1, 1.5[               | [10, 32[         | Strong       |
#' | [1.5, 2[               | [32, 100[        | Very Strong  |
#' | [2, +Inf[              | [100, +Inf[      | Decisive     |
#'
#' By default, `bfactor_log_interpret` takes (base `base`) logarithms of Bayes factors as input and returns the strength of the evidence in favor of the model/hypothesis in the numerator of the Bayes factors (usually the null hypothesis) according to the aforementioned table.
#'
#' Alternatively, and because it can be useful to consider twice the natural logarithm of the Bayes factor (which is in the same scale as the familiar deviance and likelihood ratio test statistics), \insertCite{kass1995;textual}{pcal} suggested the following scale:
#'
#' | **2*log(Bayes factor)** | **Bayes factor**      | **Evidence** |
#' | ---                     | ---                   | ---          |
#' | [-Inf, 0[               | [0, 1[                | Negative     |
#' | [0, 2[                  | [1, 3[                | Weak         |
#' | [2, 6[                  | [3, 20[               | Positive     |
#' | [6, 10[                 | [20, 150[             | Strong       |
#' | [10, +Inf[              | [150, +Inf[           | Very Strong  |
#'
#' To interpret base `base` logarithms of Bayes factors according to the latter table use `scale = "kass-raftery"`.
#'
#' When comparing Bayes factors with results from standard likelihood ratio tests it is convenient to put the null hypothesis in the denominator of the Bayes factor so that `bfactor_log_interpret` returns the strength of the evidence against the null hypothesis. If `bf` was obtained with the null hypothesis on the numerator, one can use `bfactor_log_interpret(1/bf)` or `bfactor_log_interpret(1/bf, scale = "kass-raftery")` to obtain the strength of the evidence against the null hypothesis.
#'
#' @return Returns a character vector with the same  \code{\link[base]{length}} as `bf`.
#'
#' @references
#' \insertAllCited{}
#'
#' @seealso
#' * \code{\link[pcal]{bfactor_interpret}} for the interpretation of Bayes factors in levels.
#' * \code{\link[pcal]{bfactor_to_prob}} to turn Bayes factors into posterior probabilities.
#' * \code{\link[pcal]{bcal}} for a p-value calibration that returns lower bounds on Bayes factors in favor of point null hypotheses.
#'
#' @examples
#' # Interpretation of the natural log of a Bayes factors
#' bfactor_log_interpret(log(1.5))
#' bfactor_log_interpret(log(1.5), scale = "jeffreys")
#' bfactor_log_interpret(log(1.5), scale = "kass-raftery")
#'
#' # Interpretation of the natural log of many Bayes factors
#' bfactor_log_interpret(log(c(0.1, 1.2, 3.5, 13.9, 150)))
#' bfactor_log_interpret(log(c(0.1, 1.2, 3.5, 13.9, 150)), scale = "kass-raftery")
#'
#' # Interpretation of the base-10 log of many Bayes factors
#' bfactor_log_interpret(log10(c(0.1, 1.2, 3.5, 13.9, 150)), base = 10)
#' bfactor_log_interpret(log10(c(0.1, 1.2, 3.5, 13.9, 150)), base = 10, scale = "kass-raftery")
#'
#' @export

bfactor_log_interpret <- function(bf, scale = "jeffreys", base = exp(1)) {

  .Deprecated(
    new = "bfactor_log_interpret",
    package = "polya",
    msg = "bfactor_log_interpret' is deprecated. Please use
    'polya::bfactor_log_interpret(bf)'
    instead.")

  if (is.null(bf)){
    stop(paste("Invalid argument: bf is NULL."))
  }
  if (any(!is.numeric(bf), !is.vector(bf))){
    stop(paste("Invalid argument: bf must be a numeric vector."))
  }
  if (length(bf) == 0){
    stop(paste("Invalid argument: bf is empty."))
  }
  if (all(is.na(bf))){
    stop(paste("Invalid argument: all elements of bf are NA or NaN."))
  }
  if (any(is.na(bf))){
    warning(paste("There are NA or NaN values in bf."))
  }

  if (is.null(scale)){
    stop(paste("Invalid argument: scale is NULL."))
  }
  if (any(!is.vector(scale), !is.atomic(scale), isFALSE(length(scale) == 1))){
    stop(paste("Invalid argument: scale must be an atomic vector of length 1."))
  }
  if (is.na(scale)){
    stop(paste("Invalid argument: scale is NA or NaN."))
  }
  if (!is.character(scale)){
    stop(paste("Invalid argument: the type of scale must be character."))
  }
  if (isFALSE(tolower(scale) %in% c("jeffreys", "kass-raftery"))){
    stop(paste("Invalid argument: scale must be either 'jeffreys' or 'kass-raftery'."))
  }

  if (any(
    is.null(base),
    isFALSE(is.vector(base)),
    isFALSE(is.atomic(base)),
    isFALSE(is.numeric(base)),
    isFALSE(length(base) == 1),
    is.na(base))
  ){
    stop(paste("Invalid argument: base must be a numeric vector of length 1."))
  }
  if (isTRUE(base <= 0)){
    stop(paste("Invalid argument: base must be positive."))
  }

  bf <-  base ^ bf

  switch(
    tolower(scale),
    "jeffreys" = ifelse(bf < 1, "Negative",
                        ifelse(
                          bf < 3.2, "Weak",
                          ifelse(bf < 10, "Substantial",
                                 ifelse(
                                   bf < 32, "Strong",
                                   ifelse(bf < 100, "Very Strong",
                                          "Decisive")
                                 ))
                        )),
    "kass-raftery" =   ifelse(bf < 1, "Negative",
                              ifelse(
                                bf < 3, "Weak",
                                ifelse(bf < 20, "Positive",
                                       ifelse(bf < 150, "Strong",
                                              "Very Strong"))
                              ))
  )

}

#' @title Interpretation of Bayes factors
#'
#' @description Quantify the strength of the evidence provided by the data to a model/hypothesis according a Bayes factor interpretation scale suggested by \insertCite{kass1995;textual}{pcal}.
#'
#' @param bf A numeric vector of non-negative values.
#'
#' @details Bayes factors are a summary of the evidence provided by the data to a model/hypothesis. Because it can be useful to consider twice the natural logarithm of the Bayes factor, which is in the same scale as the familiar deviance and likelihood ratio test statistics, \insertCite{kass1995;textual}{pcal} suggested the following Bayes factor interpretation scale:
#'
#' | **2*log(Bayes factor)** | **Bayes factor**      | **Evidence** |
#' | ---                     | ---                   | ---          |
#' | [-Inf, 0[               | [0, 1[                | Negative     |
#' | [0, 2[                  | [1, 3[                | Weak         |
#' | [2, 6[                  | [3, 20[               | Positive     |
#' | [6, 10[                 | [20, 150[             | Strong       |
#' | [10, +Inf[              | [150, +Inf[           | Very strong  |
#'
#' `bfactor_interpret_kr` takes Bayes factors as input and returns the strength of the evidence in favor of the model/hypothesis in the numerator of the Bayes factors (usually the null hypothesis) according to the aforementioned table.
#'
#' When comparing results with those from standard likelihood ratio tests, it is convenient to put the null hypothesis in the denominator of the Bayes factor so that `bfactor_interpret_kr` returns the strength of the evidence against the null hypothesis. If `bf` was obtained with the null hypothesis on the numerator, one can use `bfactor_interpret_kr(1/bf)` to obtain the strength of the evidence against the null hypothesis.
#'
#' @return Returns a character vector with the same \code{\link[base]{length}} as `bf`.
#'
#' @references
#' \insertAllCited{}
#'
#' @seealso
#' * \code{\link[pcal]{bfactor_interpret}} for the original interpretation scale suggested by Harold Jeffreys.
#' * \code{\link[pcal]{bfactor_log_interpret}} and \code{\link[pcal]{bfactor_log_interpret_kr}} for the interpretation of the logarithms of Bayes factors.
#' * \code{\link[pcal]{bfactor_to_prob}} to turn Bayes factors into posterior probabilities.
#' * \code{\link[pcal]{bcal}} for a p-value calibration that returns lower bounds on Bayes factors in favor of point null hypotheses.
#'
#' @examples
#' # Interpretation of one Bayes factor
#' bfactor_interpret_kr(1.5)
#'
#' # Interpretation of many Bayes factors
#' bfactor_interpret_kr(c(0.1, 1.2, 3.5, 13.9, 150))
#'
#' # Application: chi-squared goodness-of-fit test.
#' # Strength of the evidence provided by the lower
#' # bound on the Bayes factor in favor of the null hypothesis:
#' x <- matrix(c(12, 15, 14, 15), ncol = 2)
#' bfactor_interpret_kr(bcal(chisq.test(x)[["p.value"]]))
#'
#' @export

bfactor_interpret_kr <- function(bf) {

  .Deprecated(
    new = "bfactor_interpret",
    package = "polya",
    msg = "bfactor_interpret_kr' is deprecated. Please use
      'polya::bfactor_interpret(bf, scale = kass-raftery)' instead.")

  if(is.null(bf)){
    stop("Invalid argument: 'bf' is NULL.")
  }
  if(length(bf) == 0){
    stop("Invalid argument: 'bf' is empty")
  }
  if(all(is.na(bf))){
    stop("Invalid argument: all elements of 'bf' are NA or NaN.")
  }
  if(any(!is.numeric(bf), !is.vector(bf),  all(is.na(bf)))){
    stop("Invalid argument: 'bf' must be a numeric vector")
  }
  if(any(bf[!is.na(bf)] < 0)){
    stop("Invalid argument: 'bf' must be non-negative.")
  }
  if(any(is.na(bf))){
    warning("There are NA or NaN values in 'bf'.")
  }

  ifelse(bf < 1, "Negative",
         ifelse(bf < 3, "Weak",
                ifelse(bf < 20, "Positive",
                       ifelse(bf < 150, "Strong",
                              "Very Strong")
                )
         )
  )

}

#' @title Interpretation of the logarithms of Bayes factors
#'
#' @description Quantify the strength of the evidence provided by the data to a model/hypothesis according a Bayes factor interpretation scale suggested by \insertCite{kass1995;textual}{pcal}.
#'
#' @param bf A numeric vector.
#' @param base A numeric vector of \code{\link[base]{length}} one. Must be a positive number.
#'
#' @details Bayes factors are a summary of the evidence provided by the data to a model/hypothesis, and are often reported on a logarithmic scale. Because it can be useful to consider twice the natural logarithm of the Bayes factor, which is in the same scale as the familiar deviance and likelihood ratio test statistics, \insertCite{kass1995;textual}{pcal} suggested the following Bayes factor interpretation scale:
#'
#' | **2*log(Bayes factor)** | **Bayes factor**      | **Evidence** |
#' | ---                     | ---                   | ---          |
#' | [-Inf, 0[               | [0, 1[                | Negative     |
#' | [0, 2[                  | [1, 3[                | Weak         |
#' | [2, 6[                  | [3, 20[               | Positive     |
#' | [6, 10[                 | [20, 150[             | Strong       |
#' | [10, +Inf[              | [150, +Inf[           | Very strong  |
#'
#' `bfactor_log_interpret_kr` takes (base `base`) logarithms of Bayes factors as input and returns the strength of the evidence provided by the data in favor of the  model/hypothesis in the numerator of the Bayes factors (usually the null hypothesis) according to the aforementioned table.
#'
#' When comparing results with those from standard likelihood ratio tests, it is convenient to put the null hypothesis in the denominator of the Bayes factor so that `bfactor_log_interpret_kr` returns the strength of the evidence against the null hypothesis. If `bf` was obtained with the null hypothesis on the numerator, one can use `bfactor_log_interpret_kr(1/bf)` to obtain the strength of the evidence against the null hypothesis.
#'
#' @return Returns a character vector with the same {\link[base]{length}} as `bf`.
#'
#' @references
#' \insertAllCited{}
#'
#' @seealso
#' * \code{\link[pcal]{bfactor_log_interpret}} for the original interpretation scale suggested by Harold Jeffreys
#' * \code{\link[pcal]{bfactor_interpret}} and \code{\link[pcal]{bfactor_interpret_kr}} for the interpretation of Bayes factors in levels.
#' * \code{\link[pcal]{bfactor_to_prob}} to turn Bayes factors into posterior probabilities.
#' * \code{\link[pcal]{bcal}} for a p-value calibration that returns lower bounds on Bayes factors in favor of point null hypotheses.
#'
#' @examples
#' # Interpretation of one Bayes factor, natural log
#' bfactor_log_interpret_kr(1.5)
#'
#' # Interpretation of many Bayes factors, natural log
#' bfactor_log_interpret_kr(c(0.1, 1.2, 3.5, 13.9, 150))
#'
#' # Interpretation of many Bayes factors, base 10 log
#' bfactor_log_interpret_kr(log10(c(0.1, 1.2, 3.5, 13.9, 150)), base = 10)
#'
#' @export

bfactor_log_interpret_kr <- function(bf, base = exp(1)) {

  .Deprecated(
    new = "bfactor_log_interpret",
    package = "polya",
    msg = "bfactor_log_interpret_kr' is deprecated. Please use
    'polya::bfactor_log_interpret(bf, scale = kass-raftery)'
    instead.")

  if(is.null(bf)){
    stop("Invalid argument: 'bf' is NULL.")
  }
  if(length(bf) == 0){
    stop("Invalid argument: 'bf' is empty")
  }
  if(all(is.na(bf))){
    stop("Invalid argument: all elements of 'bf' are NA or NaN.")
  }
  if(any(!is.numeric(bf), !is.vector(bf),  all(is.na(bf)))){
    stop("Invalid argument: 'bf' must be a numeric vector")
  }
  if(any(is.na(bf))){
    warning("There are NA or NaN values in 'bf'.")
  }
  if(any(is.null(base), is.na(base), isFALSE(is.numeric(base)), isFALSE(is.vector(base)), isFALSE(length(base) == 1))){
    stop("Invalid argument: 'base' must be a numeric vector of length 1")
  }

  bf <-  base ^ bf

  ifelse(bf < 1, "Negative",
         ifelse(bf < 3, "Weak",
                ifelse(bf < 20, "Positive",
                       ifelse(bf < 150, "Strong",
                              "Very Strong")
                )
         )
  )

}

# nocov end





